xli.man
.\"	@(#)x11:contrib/clients/xloadimage/xli.man 1.13 94/07/29 Labtam
.\"	
.TH XLI 1 "27 Jul 1994"
.SH NAME
xli, xsetbg, xview \- load images into an X11 window or onto
the root window
.SH SYNOPSIS
\fIxli\fR [global_options] {[image_options] image ...}
.br
\fIxli\fR [global_options] [image_options] stdin < image
.SH DESCRIPTION
\fIxli\fR displays images in an X11 window or loads them onto
the root window.  See the \fIIMAGE TYPES\fR section below for
supported image types.
.PP
If the filename \fIstdin\fR is given, xli will read the image
from standard input.
.PP
If the destination display cannot support the number of colors in the
image, the image will be dithered (monochrome destination) or have its
colormap reduced (color destination) as appropriate.  This can also be
done forcibly with the \fI-halftone\fR, \fI-dither\fR, and
\fI-colors\fR options.
.PP
A variety of image manipulations can be specified, including gamma
correction, brightening, clipping, dithering, depth-reduction,
rotation, and zooming.  Most of these manipulations have simple
implementations; speed was opted for above accuracy.
.PP
If you are viewing a large image in a window, the initial window will
be at most 90% of the size of the display unless the window manager
does not correctly handle window size requests or if you've used the
\fI-fullscreen\fR or \fI-fillscreen\fR options.
You may move the image around in the window
by dragging with the first mouse button.  The cursor will indicate
which directions you may drag, if any.
.PP
When the keyboard focus is in the window you can:
.na
.nf
Type 'q' or '^C' to exit xli.
Type space, 'n' or 'f' to move to the next image in the list.
Type 'b' or 'p' to move to the previous image in the list.
Type . to reload the image.
Type l to rotate the image anti-clockwise.
Type r to rotate the image clockwise.
Type 0 to set the images assumed gamma to your display gamma
       (usually darkens images)
Type 1 to set the images assumed gamma to 1.0
       (usually lightens images)
Type 5-2 to lighten the image (5 in small steps, up to 2 in large steps)
Type 6-9 to darken the image (6 in small steps, up to 9 in large steps)
Type > resp. < to zoom in resp. out
.ad
.fi
.PP
A wide variety of common image manipulations can be done by mixing and
matching the available options.  See the section entitled \fIHINTS FOR
GOOD IMAGE DISPLAYS\fR for some ideas.
.PP
\fIXsetbg\fR is equivalent to \fIxli -onroot -quiet\fR and
\fIxview\fR is equivalent to \fIxli -view -verbose\fR.
.SH RESOURCE CLASS
\fIxli\fR uses the resource class name \fI_XSETROOT_Id\FR for
window managers which need this resource set.
.SH GLOBAL OPTIONS
The following options affect the global operation of \fIxli\fR.
They may be specified anywhere on the command line.
.TP
-default
Set the root background to the default root weave.  This is the same
as \fIxsetroot\fR with no arguments.
.TP
-debug
Talk to the X server in synchronous mode.  This is useful for
debugging.  If an X error is seen while in this mode, a core will be
dumped.
.TP
-dumpcore
Signals will not be trapped, and instead a coredump will occur.
.TP
-display \fIdisplay_name\fR
X11 display name to send the image(s) to.
.TP
-dispgamma \fIDisplay_gamma\fR
Specify the gamma correction value appropriate for the display device.
This overides the value read from the environment variable \fBDISPLAY_GAMMA\fR,
or the default value of 2.2, which is approximately correct for
many displays. A value of between 1.6 and 2.8 is reasonable. If individual
images are too bright or dark, use the -gamma option.
.PP
There is an image provided with xli called 'chkgamma.jpg' that lets you
set the display gamma reasonably accurately.  This file contains two
grayscale ramps. The ramps are chosen to look linear to the human
eye, one using continuous tones, and the other using dithering.
When the display gamma is correct, then the two ramps should look symmetrical,
and the point at which they look equally bright should be almost exactly half
way from the top to the bottom. (To find this point it helps if you move away
a little from the screen, and de-focus your eyes a bit.)
.PP
If the equal brightness point is above center increase the
gamma, and decrease it if it is below the center. The value
will usually be around 2.2 Once you've got it right, you can set
the DISPLAY_GAMMA environment variable in your .profile
.TP
-fillscreen
Use the whole screen for displaying an image. The image will be zoomed
so that it just fits the size of the screen. If -onroot is also specified,
it will be zoomed to completely fill the screen.
.TP
-fit
Force image to use the default visual and colormap.  This is useful if
you do not want technicolor effects when the colormap focus is inside
the image window, but it may reduce the quality of the displayed
image.  This is on by default if -onroot or -windowid is specified.
.TP
-fork
Fork xli.  This causes xli to disassociate itself from
the shell.  This option automatically turns on -quiet.
.TP
-fullscreen
Use the whole screen for displaying an image. The image will be surrounded by
a border if it is smaller than the screen. If -onroot is also specified,
the image will be zoomed so that it just fits the size of the screen.
.TP
-geometry \fIWxH[{+-X}{+-}Y]\fR
This sets the size of the window onto which the images are loaded to a
different value than the size of the image.  When viewing an image in
a window, this can be used to set the size and position of the
viewing window.
If the size is not specified in the geometry, (or is set to 0),
then the size will be chosen to be small enough to able to fit
the window in the screen (as usual).
.TP
-goto image_name
When the end of the list of images is reached, go to image 
\fIimage_name\fR.  This is useful for generating looped slideshows.
If more than one image of the same name as the target exists on the
argument list, the first in the argument list is used.
.TP
-help [option ...]
Give information on an option or list of options.  If no option is
given, a simple interactive help facility is invoked.
.TP
-identify
Identify the supplied images rather than display them.
.TP
-install
Forcibly install the images colormap when the window is focused.
This violates ICCCM standards and only exists to allow operation with
naive window managers.  Use this option only if your window manager
does not install colormaps properly.
.TP
-list
List the images which are along the image path.
.TP
-onroot
Load image(s) onto the root window instead of viewing in a window.
This option automatically sets the -fit option.
This is the opposite of \fI-view\fR.  \fIXSetbg\fR has this option set
by default.  If used in conjunction with -fullscreen,
the image will be zoomed to just fit. If used with -fillscreen, the image will
be zoomed to completely fill the screen. -border, -at, and -center also affect the
results.
.TP
-path
Displays the image path and image suffixes which will be used when
looking for images.  These are loaded from ~/.xlirc and
optionally from a system wide file (normally /usr/lib/xlirc).
.TP
-pixmap
Force the use of a pixmap as backing-store.  This is provided for
servers where backing-store is broken (such as some versions of the
AIXWindows server).  It may improve scrolling performance on servers
which provide backing-store.
.TP
-private
Force the use of a private colormap.  Normally colors are allocated
shared unless there are not enough colors available.
.TP
-quiet
Forces \fIxli\fR and \fIxview\fR to be quiet.  This is the
default for \fIxsetbg\fR, but the others like to whistle. 
.TP
-supported
List the supported image types. 
.TP
-verbose
Causes \fIxli\fR to be talkative, telling you what kind of
image it's playing with and any special processing that it has to do. 
This is the default for \fIxview\fR and \fIxli\fR. 
.TP
-version
Print the version number and patchlevel of this version of
\fIxli\fR.
.TP
-view
View image(s) in a window.  This is the opposite of \fI-onroot\fR and
the default for \fIxview\fR and \fIxli\fR. 
.TP
-visual \fIvisual_name\fR
Force the use of a specific visual type to display an image.  Normally
\fIxli\fR tries to pick the best available image for a
particular image type.  The available visual types are:  DirectColor,
TrueColor, PseudoColor, StaticColor, GrayScale, and StaticGray.
Nonconflicting names may be abbreviated and case is ignored.
.TP
-windowid \fIhex_window_id\fR
Sets the background pixmap of a particular window ID.  The argument
must be in hexadecimal and must be preceded by "0x" (\fIeg\fR
-windowid 0x40000b.  This is intended for setting the background
pixmap of some servers which use untagged virtual roots
(\fIeg\fR HP-VUE), but can have other interesting applications.
.SH PERSISTENT IMAGE OPTIONS
The following options may precede each image.  They take effect from
the next image, and continue until overridden or canceled with
\fI-newoptions.\fR
.TP
-border \fIcolor\fR
This sets the background portion of the window or clipped image which is
not covered by any images to be \fIcolor\fR.
.TP
-brighten \fIpercentage\fR
Specify a percentage multiplier for a color images colormap.  A value
of more than 100 will brighten an image, one of less than 100 will
darken it. 
.TP
-colors \fIn\fR
Specify the maximum number of colors to use in the image.  This is a
way to forcibly reduce the depth of an image.
.TP
-cdither
.TP
-colordither
Dither the image with a Floyd-Steinberg dither if the number of colors is reduced.
This will be slow, but will give a better looking result with a restricted color
set. \fI-cdither\fR and \fI-colordither\fR are equivalent.
.TP
-delay \fIsecs\fR
Sets xli to automatically advance to the following image,
\fIsecs\fR seconds after the next image file is displayed.
.TP
-dither
Dither a color image to monochrome using a Floyd-Steinberg dithering
algorithm.  This happens by default when viewing color images on a
monochrome display.  This is slower than \fI-halftone\fR and affects
the image accuracy but usually looks much better.
.TP
-gamma \fIImage_gamma\fR
Specify the gamma of the display the image was intended to be displayed on.
Images seem to
come in two flavors: 1) linear color images, produced by ray tracers,
scanners etc. These sort of images generally look too dark when displayed
directly to a CRT display. 2) Images that have been processed to look right
on a typical CRT display without any sort of processing. These images have
been 'gamma corrected'. By default, xli assumes that 8 bit images have
been gamma corrected and need no other processing. 24 bit images are assumed
to be linear.
If a linear image is displayed as if it is gamma corrected it will look too
dark, and a gamma value of 1.0 should be specified, so that xli can correct
the image for the CRT display device. If a gamma corrected image is displayed
as if it were a linear image, then it will look too light, and a gamma value of
(approximately) 2.2 should be specified for that image.
Some formats (RLE) allow the image gamma to be embedded as a comment in the
file itself, and the -gamma option allows overriding of the file comment.
In general, values smaller than 2.2 will lighten the image, and values
greater than 2.2 will darken the image.
In general this will work better than the -brighten option.
.TP
-gray
Convert an image to grayscale.  This is very useful when displaying
colorful images on servers with limited color capability.  The
optional spelling \fI-grey\fR may also be used.
.TP
-idelay \fIsecs\fR
Set the delay to be used for this image to \fIsecs\fR seconds (see
\fI-delay\fR).  If \fI-delay\fR was specified, this overrides it.  If
it was not specified, this sets the automatic advance delay for this
image while others will wait for the user to advance them.
.TP
-smooth
Smooth a color image.  This reduces blockiness after zooming an image
up.  If used on a monochrome image, nothing happens.  This option can
take awhile to perform, especially on large images.  You may specify
more than one \fI-smooth\fR option per image, causing multiple
iterations of the smoothing algorithm.
.TP
-title \fIwindow_title\fR
Set the titlebar of the window used to display the image.
This will overide any title that is read from the image
file. The title will also be used for the icon name.
.TP
-xpm \fIcolor_context_key\fR
Select the prefered xpm colour map. XPM files may contain more than one
color mapping, each mapping being appropriate for a particular visual.
Normally xli will select an apropriate color mapping from that supported
by the XPM file by checking on the default X visual class and depth.
This option allows the user to overide this choice.
Legal values of  \fIcolor_context_key\fR are: m, g4, g and c.
m = mono, g4 = 4 level gray, g = gray, c = color ).
.TP
-xzoom \fIpercentage\fR
Zoom the X axis of an image by \fIpercentage\fR.  A number greater
than 100 will expand the image, one smaller will compress it.  A zero
value will be ignored.  This option, and the related \fI-yzoom\fR are
useful for correcting the aspect ratio of images to be displayed.
.TP
-yzoom \fIpercentage\fR
Zoom the Y axis of an image by \fIpercentage\fR.  See \fI-xzoom\fR for
more information. 
.TP
-zoom \fIpercentage\fR
Zoom both the X and Y axes by \fIpercentage\fR.  See \fI-xzoom\fR for
more information.  Technically the percentage actually zoomed is the
square of the number supplied since the zoom is to both axes, but I
opted for consistency instead of accuracy.
.TP
\-zoom auto
Zoom large images to fit the screen; don't zoom small images.
.TP
-newoptions
Reset options that propagate.  The \fI-bright, -colors, -colordither, -delay,
-dither, -gamma, -gray, -normalize, -smooth, -xzoom, -yzoom\fR, and
\fI-zoom\fR options normally propagate to all following images.
.SH LOCAL IMAGE OPTIONS
The following options may precede each image.  These options are
local to the image they precede. 
.TP
-at \fIX\fR,\fIY\fR
Indicates coordinates to load the image at \fIX\fR,\fIY\fR on the base image.
If this is an option to the first image, and the \fI-onroot\fR option is
specified, the image will be loaded at the given location on the
display background. 
.TP
-background \fIcolor\fR
Use \fIcolor\fR as the background color instead of the default
(usually white but this depends on the image type) if you are
transferring a monochrome image to a color display. 
.TP
-center
Center the image on the base image loaded.  If this is an option to
the first image, and the \fI-onroot\fR option is specified, the image
will be centered on the display background. 
.TP
-clip \fIX\fR,\fIY\fR,\fIW\fR,\fIH\fR
Clip the image before loading it.  \fIX\fR and \fIY\fR define the
upper-left corner of the clip area, and \fIW\fR and \fIH\fR define the
extents of the area.  A zero value for \fIW\fR or \fIH\fR will be
interpreted as the remainder of the image. 
Note that \fIX\fR and \fIY\fR may be negative, and that \fIW\fR and \fIH\fR
may be larger than the image. This causes a border to be placed around the
image. The border color may be set with the \fI-border\fR option.
.TP
-expand
Forces the image (after all other optional processing) to be expanded
into a True Color (24 bit) image. This is useful on systems which support
24 bit color, but where xli might choose to load a bitmap or 8 bit image
into one of the other smaller depth visuals supported on your system.
.TP
-foreground \fIcolor\fR
Use \fIcolor\fR as the foreground color instead of black if you are
transferring a monochrome image to a color display.  This can also be
used to invert the foreground and background colors of a monochrome
image. 
.TP
-halftone
Force halftone dithering of a color image when displaying on a
monochrome display.  This option is ignored on monochrome images.
This dithering algorithm blows an image up by sixteen times; if you
don't like this, the \fI-dither\fR option will not blow the image up
but will take longer to process and will be less accurate.
.TP
-invert
Inverts a monochrome image.  This is shorthand for \fI-foreground
white -background black\fR.
.TP
-merge
Merge this image onto the base image after local processing.  The base
image is considered to be the first image specified or the last image
that was not preceded by \fI-merge\fR.  If used in conjunction with
\fI-at\fR and \fI-clip\fR, very complex images can be built up.
Note that the final image will be the size of the first image, and that
subsequent merged images overlay previous images. The final image size
can be altered by using the \fI-clip\fR option on the base image to
make it bigger or smaller.
This option is on by default for all images if the \fI-onroot\fR or
\fI-windowid\fR options are specified.
.TP
-name \fIimage_name\fR
Force the next argument to be treated as an image name.  This is
useful if the name of the image is \fI-dither\fR, for instance. 
.TP
-normalize
Normalize a color image.
.TP
-rotate \fIdegrees\fR
Rotate the image by \fIdegrees\fR clockwise.  The number must be a
multiple of 90.
.SH EXAMPLES
To load the rasterfile "my.image" onto the background and replicate
it to fill the entire background:
.sp
.ti +5
xli -onroot my.image
.PP
To load a monochrome image "my.image" onto the background, using red
as the foreground color, replicate the image, and overlay
"another.image" onto it at coordinate (10,10):
.sp
.ti +5
xli -foreground red my.image -at 10,10 another.image
.PP
To center the rectangular region from 10 to 110 along the X axis and
from 10 to the height of the image along the Y axis:
.sp
.ti +5
xli -center -clip 10,10,100,0 my.image
.PP
To double the size of an image:
.sp
.ti +5
xli -zoom 200 my.image
.PP
To halve the size of an image:
.sp
.ti +5
xli -zoom 50 my.image
.PP
To brighten a dark image:
.sp
.ti +5
xli -brighten 150 my.image
.PP
To darken a bright image:
.sp
.ti +5
xli -brighten 50 my.image
.SH HINTS FOR GOOD IMAGE DISPLAYS
Since images are likely to come from a variety of sources, they may be
in a variety of aspect ratios which may not be supported by your
display.  The \fI-xzoom\fR and \fI-yzoom\fR options can be used to
change the aspect ratio of an image before display.  If you use these
options, it is recommended that you increase the size of one of the
dimensions instead of shrinking the other, since shrinking looses
detail.  For instance, many GIF and G3 FAX images have an X:Y ratio of
about 2:1.  You can correct this for viewing on a 1:1 display with
either \fI-xzoom 50\fR or \fI-yzoom 200\fR (reduce X axis to 50% of
its size and expand Y axis to 200% of its size, respectively) but the
latter should be used so no detail is lost in the conversion.
.PP
When zooming color images up you can reduce blockiness with
\fI-smooth\fR.  For zooms of 300% or more, I recommend two smoothing
passes (although this can take awhile to do on slow machines).  There
will be a noticeable improvement in the image.
.PP
You can perform image processing on a small portion of an image by
loading the image more than once and using the \fI-merge\fR, \fI-at\fR
and \fI-clip\fR options.  Load the image, then merge it with a
clipped, processed version of itself.  To brighten a 100x100 rectangular
portion of an image located at (50,50), for instance, you could type:
.sp
.ti +5
xli my.image -merge -at 50,50 -clip 50,50,100,100 -brighten 150 my.image
.PP
If you're using a display with a small colormap to display colorful
images, try using the \fI-gray\fR option to convert to grayscale.
.SH XLITO
\fIxlito\fR (XLoadImageTrailingOptions) is a separate utility that provides
a file format independent way of marking image files with the appropriate

CHANGES
Fri Nov 10 21:50:45 PST 2006

Fix crashes for PGM/PPM images with maxval > 255.
Fix various uninitialized variable warnings.

Sun Sep  4 21:59:17 PDT 2005

Finally fix the unsigned arithmetic problem that was resulting in blank screens
when using -fillscreen with -onroot.  Also fix some problems turned up with
-Wmissing-prototypes.

Sun Feb 27 15:16:08 PST 2005

Fix a security problem in the faces loader, a security problem when
opening compressed files, and check for integer overflows in image data
size calculations.

Add support for bzip2-compressed files.

#include <assert.h> in png.c, not "assert.h" -- no idea why I did that.

Sat Mar 29 16:56:37 PST 2003

Use integer arithmetic in zoom calculuations instead of floating point.

Remove unused varargs.h #include from rlelib.c

Fix bug when merging more than two images.

Mon May 21 09:43:49 PDT 2001

Don't abort loading GIF87a images if the screen descriptor is non-null.
Emit a warning message and continue.

Sun Jul 30 00:11:34 PDT 2000

skip past failed loads when going backwards

Thu Feb 17 06:33:10 PST 2000

fix problem where -geometry +0+0 gets ignored


xli.c
/*
 * xli.c:
 *
 * generic image loader for X11
 *
 * smar@reptiles.org 1998/09/30
 *
 * Graeme Gill 93/2/24
 *
 * Based on xloadimage by
 * jim frost 09.27.89
 *
 * Copyright 1989, 1990, 1991 Jim Frost.
 * See included file "copyright.h" for complete copyright information.
 */

#include "copyright.h"
#include "xli.h"
#include "patchlevel"
#include <signal.h>

GlobalsRec globals;

/*
 * used for the -default option.  this is the root weave bitmap with
 * the bits in the order that xli likes.
 */

#define root_weave_width 4
#define root_weave_height 4
static byte root_weave_bits[] =
{
	0xe0, 0xb0, 0xd0, 0x70
};

#define INIT_IMAGE_OPTS(istr){	\
    istr.name = (char *) 0;	\
    istr.fullname = (char *) 0;	\
    istr.loader_idx = -1;	\
    istr.atx = istr.aty = 0;	\
    istr.ats = FALSE;		\
    istr.border = (char *) 0;	\
    istr.bright = 0;		\
    istr.center = FALSE;	\
    istr.clipx = istr.clipy = 0;\
    istr.clipw = istr.cliph = 0;\
    istr.colordither = FALSE;	\
    istr.colors = 0;		\
    istr.delay = -1;		\
    istr.dither = 0;		\
    istr.gamma = UNSET_GAMMA;	\
    istr.gray = FALSE;		\
    istr.iscale = 0;		\
    istr.iscale_auto = FALSE;	\
    istr.merge = FALSE;		\
    istr.normalize = FALSE;	\
    istr.rotate = 0;		\
    istr.smooth = FALSE;	\
    istr.title = NULL;		\
    istr.xpmkeyc = 0;		\
    istr.xzoom = istr.yzoom = 0;\
    istr.fg = (char *) 0;	\
    istr.bg = (char *) 0;	\
    istr.done_to = 0;	\
    istr.to_argv = 0; }

/* deal with unsigned arithmetic problems */
static int scentre(unsigned tspan, unsigned sspan)
{
	return ((int) tspan - (int) sspan) / 2;
}

int main(int argc, char *argv[])
{
	Image *idisp;
	ImageOptions *images;
	int maximages, nimages, i, first, dir;
	unsigned int winwidth, winheight;
	ImageOptions persist_ops;
	char switchval;

	CURRFUNC("main");

	maximages = INIT_MAXIMAGES;
	images = (ImageOptions *) lmalloc(maximages * sizeof(ImageOptions));

	persist_ops.border = (char *) 0;
	persist_ops.bright = 0;
	persist_ops.colordither = FALSE;
	persist_ops.colors = 0;		/* remaining images */
	persist_ops.delay = -1;
	persist_ops.dither = 0;
	persist_ops.gamma = UNSET_GAMMA;
	persist_ops.normalize = FALSE;
	persist_ops.smooth = FALSE;
	persist_ops.xpmkeyc = 0;	/* none */
	persist_ops.xzoom = 0;
	persist_ops.yzoom = 0;
	persist_ops.zoom_auto = FALSE;
	persist_ops.iscale = 0;
	persist_ops.iscale_auto = FALSE;

	/* set up internal error handlers */
	signal(SIGSEGV, internalError);
	signal(SIGBUS, internalError);
	signal(SIGFPE, internalError);
	signal(SIGILL, internalError);

#if defined(_AIX) && defined(_IBMR2)
	/*
	 * the RS/6000 (AIX 3.1) has a new signal, SIGDANGER, which you get
	 * when memory is exhausted.  since malloc() can overcommit,
	 * it's a good idea to trap this one.
	 */
	signal(SIGDANGER, memoryExhausted);
#endif

	if (argc < 2)
		usage(argv[0]);

	/*
	 * defaults and other initial settings.  some of these depend on what
	 * our name was when invoked.
	 */

	loadPathsAndExts();

	xliDefaultDispinfo(&globals.dinfo);
	globals.argv0 = argv[0];	/* so we can get at this elsewhere */
	globals._Xdebug = FALSE;
	globals._DumpCore = FALSE;
	globals.onroot = FALSE;
	globals.verbose = TRUE;
	if (!strcmp(tail(argv[0]), "xview")) {
		globals.onroot = FALSE;
		globals.verbose = TRUE;
	} else if (!strcmp(tail(argv[0]), "xsetbg")) {
		globals.onroot = TRUE;
		globals.verbose = FALSE;
	}

	/* Get display gamma from the environment if we can */
	{
		char *gstr;

		globals.display_gamma = UNSET_GAMMA;
		if ((gstr = getenv("DISPLAY_GAMMA")) != NULL) {
			globals.display_gamma = atof(gstr);
		}

		/* Ignore silly values */
		if (globals.display_gamma > 20.0 || globals.display_gamma < 0.05)
			globals.display_gamma = DEFAULT_DISPLAY_GAMMA;
	}

	globals.dname = NULL;
	globals.fit = FALSE;
	globals.forall = FALSE;
	globals.fillscreen = FALSE;
	globals.fullscreen = FALSE;
	globals.go_to = NULL;
	globals.do_fork = FALSE;
	globals.identify = FALSE;
	globals.install = FALSE;
	globals.private_cmap = FALSE;
	globals.use_pixmap = FALSE;
	globals.set_default = FALSE;
	globals.user_geometry = NULL;
	globals.visual_class = -1;
	winwidth = winheight = 0;

	nimages = 0;
	INIT_IMAGE_OPTS(images[nimages]);
	for (i = 1; i < argc; i++) {
		OptionId opid;

		opid = optionNumber(argv[i]);
		if (opid == OPT_BADOPT) {
			printf("%s: Bad option\n", argv[i]);
			usage(argv[0]);
		}

		if (opid == OPT_SHORTOPT) {
			printf("%s: Not enough characters to identify option\n", argv[i]);
			usage(argv[0]);
		}

		if (opid == OPT_NOTOPT || opid == NAME) {
			if (opid == NAME && argv[++i] == NULL)
				continue;

			if (nimages + 1 >= maximages) {
				maximages *= 2;
				images = (ImageOptions *) lrealloc(
					(byte *) images,
					maximages * sizeof(ImageOptions));
			}

			images[nimages].name = argv[i];
			images[nimages].border = persist_ops.border;
			images[nimages].bright = persist_ops.bright;
			images[nimages].colordither = persist_ops.colordither;
			images[nimages].colors = persist_ops.colors;
			images[nimages].delay = persist_ops.delay;
			images[nimages].dither = persist_ops.dither;
			images[nimages].gamma = persist_ops.gamma;
			images[nimages].normalize = persist_ops.normalize;
			images[nimages].smooth = persist_ops.smooth;
			images[nimages].xpmkeyc = persist_ops.xpmkeyc;
			images[nimages].xzoom = persist_ops.xzoom;
			images[nimages].yzoom = persist_ops.yzoom;
			images[nimages].zoom_auto = persist_ops.zoom_auto;
			images[nimages].iscale_auto = persist_ops.iscale_auto;
			nimages++;
			INIT_IMAGE_OPTS(images[nimages]);
		} else if (isGeneralOption(opid)) {
			i += doGeneralOption(opid, &argv[i], &persist_ops,
				&images[nimages]);
		} else if (isLocalOption(opid)) {
			i += doLocalOption(opid, &argv[i], TRUE, &persist_ops,
				&images[nimages]);
		} else {
			printf("%s: Internal error parsing arguments\n",
				argv[0]);
			exit(1);
		}
	}

	if (globals._DumpCore) {
		signal(SIGSEGV, SIG_DFL);
		signal(SIGBUS, SIG_DFL);
		signal(SIGFPE, SIG_DFL);
		signal(SIGILL, SIG_DFL);
	}

	if (globals.fit && (globals.visual_class != -1)) {
		printf("-fit and -visual options are mutually exclusive (ignoring -visual)\n");
		globals.visual_class = -1;
	}
	if (!nimages && !globals.set_default)
		exit(0);

	if (globals.identify) {	/* identify the named image(s) */
		for (i = 0; i < nimages; i++)
			identifyImage(images[i].name);
		exit(0);
	}

	/* start talking to the display */
	if (!xliOpenDisplay(&globals.dinfo, globals.dname)) {
		fprintf(stderr, "%s: Cannot open display '%s'\n",
			globals.argv0, xliDisplayName(globals.dname));
		exit(1);
	}

	if (globals.do_fork) {
		switch (fork()) {
		case -1:
			perror("fork");
			/* FALLTHRU */
		case 0:
			break;
		default:
			exit(0);
		}
	}

	/* -default: resets colormap and load default root weave */

	if (globals.set_default) {
		byte *old_data;
		ImageOptions ioptions;
		boolean tempverb;

		INIT_IMAGE_OPTS(ioptions);
		idisp = newBitImage(root_weave_width, root_weave_height);
		idisp->gamma = globals.display_gamma;
		old_data = idisp->data;
		idisp->data = root_weave_bits;
		tempverb = globals.verbose;
		globals.verbose = FALSE;
		imageOnRoot(&globals.dinfo, idisp, &ioptions);
		globals.verbose = tempverb;
		idisp->data = old_data;
		freeImage(idisp);
		if (!nimages) {
			xliCloseDisplay(&globals.dinfo);
			exit(0);
		}
	}

	/* load in each named image */
	idisp = 0;
	first = -1;
	dir = 1;
	for (i = 0; i < nimages; i += dir) {
		ImageOptions *io;
		Image *inew, *itmp;

		if (i < 0) {
			dir = 1;
			continue;
		}

		io = &images[i];
		if (io->iscale) {
			io->xzoom = io->yzoom = io->iscale < 0 ? 
				100 << -io->iscale : 100 >> io->iscale;
		}

		if (!(inew = loadImage(io, globals.verbose)))
			continue;

		first = (first < 0);

		if (inew->flags & FLAG_ISCALE)
			io->xzoom = io->yzoom = 0;

		if (io->gamma != UNSET_GAMMA)
			inew->gamma = io->gamma;
		else if (UNSET_GAMMA == inew->gamma)
			defaultgamma(inew);

		/* Process any other options that need doing here */
		if (io->border) {
			xliParseXColor(&globals.dinfo, io->border,
				&io->bordercol);
			xliGammaCorrectXColor(&io->bordercol,
				DEFAULT_DISPLAY_GAMMA);
		}

		/*
		 * if first image and we're putting it on the root window
		 * in fullscreen mode, set zoom factors to something reasonable
		 */

		if ( (first || globals.forall) && 
			      ((globals.onroot && globals.fullscreen) ||
				globals.fillscreen) && !io->xzoom &&
				!io->yzoom && !io->center) {
			double wr, hr;

			wr = (double) globals.dinfo.width / inew->width;
			hr = (double) globals.dinfo.height / inew->height;
			io->xzoom = io->yzoom = (((wr < hr) ^
				(globals.onroot && globals.fillscreen)) ?
				wr : hr) * 100 + 0.5;
		}

		itmp = processImage(&globals.dinfo, inew, io);
		if (itmp != inew)
			freeImage(inew);

		inew = itmp;

		if (idisp) {
			if (io->center) {
				io->atx = scentre(idisp->width, inew->width);
				io->aty = scentre(idisp->height, inew->height);
			}
			if (!idisp->title) {
				idisp->title = dupString(inew->title);
				idisp->gamma = inew->gamma;
			}
			itmp = merge(idisp, inew, io->atx, io->aty, io);
			if (idisp != itmp) {
				freeImage(idisp);
				idisp = itmp;
			}
			freeImage(inew);
			inew = idisp;
		}

		/*
		 * else if this is the first image on root
		 * set its position and any border needed
		 */
		if ( (first || globals.forall) && 
				globals.onroot && (winwidth || winheight ||
				io->center || io->ats || globals.fullscreen ||
				globals.fillscreen)) {
			int atx = io->atx, aty = io->aty;

			if (!winwidth)
				winwidth = globals.dinfo.width;
			if (!winheight)
				winheight = globals.dinfo.height;

			if (!images[0].ats) {
				atx = scentre(winwidth, inew->width);
				aty = scentre(winheight, inew->height);
			}
			/* use clip to put border around image */
			itmp = clip(inew, -atx, -aty, winwidth, winheight, io);
			if (itmp != inew) {
				freeImage(inew);
				inew = itmp;
			}
			idisp = inew;
		} else if (!idisp) {
			idisp = inew;
		}

		/* if next image is to be merged onto this one,
		 * then go on to the next image.
		 */

		if (globals.onroot || ((i + 1 < nimages) &&
				(images[i + 1].merge)))
			continue;

		dir = 1;
		switchval = imageInWindow(&globals.dinfo, idisp, io,
			argc, argv);

		switch (switchval) {
		/* window got nuked by someone */
		case '\0':
			xliCloseDisplay(&globals.dinfo);
			exit(1);

		/* user quit */
		case '\003':
		case 'q':
			cleanUpWindow(&globals.dinfo);
			xliCloseDisplay(&globals.dinfo);
			exit(0);

		/* delete current image */
		case 'x':
			/* double-check delete-enable here */
			if (globals.delete && io->fullname) {
				if (unlink(io->fullname) < 0) {
					perror(io->fullname);
				} else {
					fprintf(stderr, "Deleted %s\n",
						io->fullname);
				}
			} else {
				i--;
			}

		/* next image */
		case ' ':
		case 'f':
		case 'n':
			if (i >= (nimages - 1) && globals.go_to != NULL) {
				int j;

				for (j = 0; j < nimages; j++) {
					if (!strcmp(images[j].name,
							globals.go_to)) {
						i = j - 1;
						break;
					}
				}

				if (j >= nimages)
					fprintf(stderr, "Target for -goto %s was not found\n", globals.go_to);
			}
			break;

		/* previous image */
		case 'b':
		case 'p':
			dir = -1;
			break;

		case '.':	/* re-load current image */
			i--;
			break;

		case '0':
		case '1':
		case '2':
		case '3':
		case '4':
		case '5':
		case '6':
		case '7':
		case '8':
		case '9':
			/* change gamma same way as a command line option */
			if (UNSET_GAMMA == io->gamma)
				io->gamma = idisp->gamma;

			switch (switchval) {
				int l;
				double jump;

			/* set gamma to default image gamma */
			case '0':	
				io->gamma = DEFAULT_DISPLAY_GAMMA;
				break;

			/* set gamma to 1.0 */
			case '1':	
				io->gamma = 1.0;
				break;

			default:
				l = (switchval - '6');
				if (l >= 0)
					l++;
				if (l > 0)
					jump = 1.1;
				else
					jump = 0.909;
				if (l < 0)
					l = -l;
				for (l--; l > 0; l--)
					jump *= jump;
				io->gamma *= jump;
				break;
			}
			if (globals.verbose)
				printf("Image Gamma now %f\n", io->gamma);
			i--;
			break;

		/* rotations */
		case 'l':
		case 'r':
			switch (switchval) {
			case 'l':
				io->rotate -= 90;
				break;
			case 'r':
				io->rotate += 90;
				break;
			}
			while (io->rotate >= 360)
				io->rotate -= 360;
			while (io->rotate < 0)
				io->rotate += 360;
			if ( globals.forall && 
					((globals.onroot && globals.fullscreen) ||
					 globals.fillscreen) && !io->center) {
				double wr, hr,f;
				wr = (double) globals.dinfo.width / inew->height;
				hr = (double) globals.dinfo.height / inew->width;
				f = (((wr < hr) ^ (globals.onroot && globals.fillscreen)) ?  wr : hr) * 100;
				io->xzoom *= f/100.0;
				io->yzoom *= f/100.0;
			}
			if (globals.verbose)
				printf("Image rotation is now %d\n",
					io->rotate);
			i--;
			break;

		case '<':
		case '=':
		case '>':
			switch (switchval) {
			case '<':
				io->iscale += 1;
				break;
	
			case '>':
				io->iscale -= 1;
				break;

			case '=':
				io->iscale = 0;
				break;
			}
			io->xzoom = io->yzoom = 0;
			io->zoom_auto = 0;
			io->iscale_auto = 0;

			if (globals.verbose) {
				printf("Image decoder scaling is now %d\n",
					io->iscale);
			}
			i--;
			break;
		}
		freeImage(idisp);
		idisp = 0;
	}

	if (idisp && globals.onroot)
		imageOnRoot(&globals.dinfo, idisp, &images[nimages - 1]);
	xliCloseDisplay(&globals.dinfo);
	exit(0);
}
